components/script/dom/htmlimageelement.rs

```diff
@@ -356,20 +390,244 @@ impl HTMLImageElement {
     }
 
     /// <https://html.spec.whatwg.org/multipage/#update-the-source-set>
-    fn update_source_set(&self) -> Vec<DOMString> {
+    fn update_source_set(&self) {
+        // Step 1
+        *self.source_set.borrow_mut() = HTMLImageElement::new_source_set();
+
+        // Step 2
         let elem = self.upcast::<Element>();
-        // TODO: follow the algorithm
-        let src = elem.get_string_attribute(&local_name!("src"));
-        if src.is_empty() {
-            return vec![]
+        let parent = elem.upcast::<Node>().GetParentElement();
+        let is_parent_picture = match parent.clone() {
+            Some(xparent) => xparent.is::<HTMLPictureElement>(),
+            None => false
+        };
+        let nodes = parent.unwrap().upcast::<Node>().ChildNodes();
+        let elements = match is_parent_picture {
+            true =>
+            nodes.iter().filter_map(DomRoot::downcast::<Element>)
+                        .map(|n| DomRoot::from_ref(&*n)).collect(),
+            false => vec![DomRoot::from_ref(&*elem)],
+        };
+
+        // Step 3
+        let _width = match elem.get_attribute(&ns!(), &local_name!("width")) {
```

Comment:
We should be using this value.

Reply:
Actually it was supposed to be used in `parse_a_sizes_attribute` as a fallback. But since we removed the last few steps of this algorithm on account of `evaluate` handling it for us, it is not getting used now.